ee/hogai/graph/agent_modes/nodes.py

```diff
@@ -17,67 +16,137 @@
     ReplaceMessages,
 )
-from ee.hogai.utils.types.composed import MaxNodeName
+from ee.hogai.utils.types.base import NodePath
 
+from .compaction_manager import AnthropicConversationCompactionManager
 from .prompts import (
+    AGENT_PROMPT,
+    BASIC_FUNCTIONALITY_PROMPT,
+    CORE_MEMORY_INSTRUCTIONS_PROMPT,
+    DOING_TASKS_PROMPT,
+    PROACTIVENESS_PROMPT,
+    ROLE_PROMPT,
     ROOT_BILLING_CONTEXT_ERROR_PROMPT,
     ROOT_BILLING_CONTEXT_WITH_ACCESS_PROMPT,
     ROOT_BILLING_CONTEXT_WITH_NO_ACCESS_PROMPT,
     ROOT_CONVERSATION_SUMMARY_PROMPT,
     ROOT_GROUPS_PROMPT,
     ROOT_HARD_LIMIT_REACHED_PROMPT,
-    ROOT_SYSTEM_PROMPT,
     ROOT_TOOL_DOES_NOT_EXIST,
-)
-from .tools import (
-    CreateAndQueryInsightTool,
-    CreateDashboardTool,
-    ReadDataTool,
-    ReadTaxonomyTool,
-    SearchTool,
-    SessionSummarizationTool,
-    TodoWriteTool,
+    TASK_MANAGEMENT_PROMPT,
+    TONE_AND_STYLE_PROMPT,
+    TOOL_USAGE_POLICY_PROMPT,
+    WRITING_STYLE_PROMPT,
 )
 
 if TYPE_CHECKING:
     from ee.hogai.tool import MaxTool
 
-SLASH_COMMAND_INIT = "/init"
-SLASH_COMMAND_REMEMBER = "/remember"
-
 
 RootMessageUnion = HumanMessage | AssistantMessage | FailureMessage | AssistantToolCallMessage | ContextMessage
 T = TypeVar("T", RootMessageUnion, BaseMessage)
 
-RootTool = Union[type[BaseModel], "MaxTool"]
+logger = structlog.get_logger(__name__)
+
+DEFAULT_TOOLS: list[type["MaxTool"]] = [
+    ReadTaxonomyTool,
+    ReadDataTool,
+    SearchTool,
+    TodoWriteTool,
+]
 
 logger = structlog.get_logger(__name__)
 
 
-class RootNode(AssistantNode):
+class AgentToolkit:
+    def __init__(self, *, team: Team, user: User, context_manager: AssistantContextManager):
+        self._team = team
+        self._user = user
+        self._context_manager = context_manager
+
+    @property
+    def default_tools(self) -> list[type["MaxTool"]]:
+        return DEFAULT_TOOLS.copy()
+
+    @property
+    def custom_tools(self) -> list[type["MaxTool"]]:
+        """
+        Custom tools are tools that are not part of the default toolkit.
+        """
+        return []
+
+    async def get_tools(self, state: AssistantState, config: RunnableConfig) -> list["MaxTool"]:
+        from ee.hogai.registry import get_contextual_tool_class
+
+        tool_classes: list[type[MaxTool]] = [*self.default_tools, *self.custom_tools]
+
+        # Processed tools
+        available_tools: list[MaxTool] = []
+
+        # Initialize the static toolkit
+        dynamic_tools = (
+            tool_class.create_tool_class(
+                team=self._team,
+                user=self._user,
+                state=state,
+                config=config,
+                context_manager=self._context_manager,
+            )
+            for tool_class in tool_classes
+        )
+        available_tools.extend(await asyncio.gather(*dynamic_tools))
+
+        # Inject contextual tools
+        tool_names = self._context_manager.get_contextual_tools().keys()
```

Comment:
Idea for the future: contextual tools (basically frontend tools) should define which mode they're attached to, e.g. the surveys frontend tools should be in the surveys mode.